solr/core/src/java/org/apache/solr/core/SolrCore.java

```diff
@@ -3236,6 +3238,14 @@ public void postClose(SolrCore core) {
     });
     return blobRef;
   }
+  public ShardStateProvider getShardStateProvider(){
+    if(this.shardStateProvider == null) {
+      if(coreContainer.getZkController() != null){
+        this.shardStateProvider = coreContainer.getZkController().getZkStateReader().getShardStateProvider(this.getCoreDescriptor().getCloudDescriptor().getCollectionName());
```

Comment:
High level, are there any concerns with mismatched state between the data provided by `ShardStateProvider` and the other ZK data that is read among all of these flows?

I know that in the past we've had to commit a few fixes due to mismatched zk states (but that was due to multiple zk readers last time).